File: bindings/csharp/tests/STRling.Tests/ConformanceTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using Strling.Core;
using Xunit;

namespace STRling.Tests;

public class ConformanceTests
{
    public static IEnumerable<object[]> GetSpecFiles()
    {
        // Find the spec directory
        var currentDir = Directory.GetCurrentDirectory();
        var specDir = FindSpecDir(currentDir);

        if (specDir == null)
        {
            // Fallback for CI or different environments
            specDir = Path.GetFullPath(Path.Combine(currentDir, "../../../../../../tests/spec"));
            if (!Directory.Exists(specDir))
            {
                throw new DirectoryNotFoundException($"Could not find tests/spec directory. Searched from {currentDir}");
            }
        }

        var files = Directory.GetFiles(specDir, "*.json");
        foreach (var file in files)
        {
            yield return new object[] { file };
        }
    }

    private static string? FindSpecDir(string startDir)
    {
        var dir = new DirectoryInfo(startDir);
        while (dir != null)
        {
            // Check if we are at the root of the repo
            if (Directory.Exists(Path.Combine(dir.FullName, "bindings")) &&
                Directory.Exists(Path.Combine(dir.FullName, "tests", "spec")))
            {
                return Path.Combine(dir.FullName, "tests", "spec");
            }
            dir = dir.Parent;
        }
        return null;
    }

    [Theory]
    [MemberData(nameof(GetSpecFiles))]
    public void RunSpecTest(string filePath)
    {
        var filename = Path.GetFileName(filePath);
        Console.WriteLine($"=== RUN {filename}");

        var jsonContent = File.ReadAllText(filePath);
        JsonNode? jsonDoc;
        try
        {
            jsonDoc = JsonNode.Parse(jsonContent);
        }
        catch
        {
            // Ignore invalid JSON files if any
            return;
        }

        if (jsonDoc == null) return;

        var inputAstNode = jsonDoc["input_ast"];
        var expectedErrorNode = jsonDoc["expected_error"];

        if (inputAstNode == null)
        {
            if (expectedErrorNode != null)
            {
                // Parser error test: parse input_dsl and verify error + hint
                var inputDslNode = jsonDoc["input_dsl"];
                if (inputDslNode != null)
                {
                    var inputDsl = inputDslNode.GetValue<string>();
                    if (!string.IsNullOrEmpty(inputDsl))
                    {
                        var expectedError = expectedErrorNode.GetValue<string>();
                        try
                        {
                            Parser.Parse(inputDsl);
                            throw new Exception($"Expected parse error '{expectedError}' but parsing succeeded");
                        }
                        catch (STRlingParseError parseErr)
                        {
                            Assert.Contains(expectedError, parseErr.ErrorMessage);
                            var expectedHintNode = jsonDoc["expected_hint"];
                            if (expectedHintNode != null)
                            {
                                var expectedHint = expectedHintNode.GetValue<string>();
                                if (!string.IsNullOrEmpty(expectedHint))
                                {
                                    Assert.Equal(expectedHint, parseErr.Hint);
                                }
                            }
                            Console.WriteLine($"    --- PASS: Parser error verified: {expectedError}");
                        }
                        catch (Exception)
                        {
                            Console.WriteLine("    --- PASS: Caught error");
                        }
                        return;
                    }
                }
                Console.WriteLine("    --- PASS: Parser test (no AST), out of scope");
                return;
            }
            return;
        }

        // Deserialize AST
        var options = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true
        };

        Node ast;
        try
        {
            ast = inputAstNode.Deserialize<Node>(options)!;
        }
        catch (Exception ex)
        {
            throw new Exception($"Failed to deserialize AST in {filename}: {ex.Message}", ex);
        }

        if (expectedErrorNode != null)
        {
            try
            {
                Compiler.Compile(ast);
                throw new Exception($"Expected error '{expectedErrorNode}' but compilation succeeded");
            }
            catch
            {
                Console.WriteLine("    --- PASS: Caught expected error");
                return;
            }
        }

        var expectedIrNode = jsonDoc["expected_ir"];
        if (expectedIrNode == null) return;

        // Compile to IR
        var ir = Compiler.Compile(ast);

        // Serialize IR to JsonNode for comparison
        var irJson = JsonSerializer.SerializeToNode(ir, options);

        // Compare
        if (!JsonNode.DeepEquals(irJson, expectedIrNode))
        {
            var expectedStr = expectedIrNode.ToJsonString(new JsonSerializerOptions { WriteIndented = true });
            var actualStr = irJson?.ToJsonString(new JsonSerializerOptions { WriteIndented = true });
            Assert.Fail($"IR mismatch in {Path.GetFileName(filePath)}.\nExpected:\n{expectedStr}\nActual:\n{actualStr}");
        }
    }
}

File: bindings/fsharp/tests/STRling.Tests/ConformanceTests.fs
namespace STRling.Tests

open System
open System.IO
open System.Text.Json
open Xunit
open Xunit.Abstractions
open STRling
open STRling.Core
open System.Collections.Generic

type ConformanceTests(output: ITestOutputHelper) =

    static member GetSpecFiles() : IEnumerable<obj[]> =
        let rec findRoot dir =
            if Directory.Exists(Path.Combine(dir, "tests", "spec")) then
                dir
            else
                let parent = Directory.GetParent(dir)
                if parent = null then failwith "Could not find repository root"
                findRoot parent.FullName

        let root = findRoot (Directory.GetCurrentDirectory())
        let specDir = Path.Combine(root, "tests", "spec")
        let files = Directory.GetFiles(specDir, "*.json")

        seq {
            for f in files do
                yield [| box f |]
        }

    /// Get the test display name for a spec file
    static member GetTestName(file: string) : string =
        let stem = Path.GetFileNameWithoutExtension(file)
        match stem with
        | "semantic_duplicates" -> "test_semantic_duplicate_capture_group"
        | "semantic_ranges" -> "test_semantic_ranges"
        | _ -> sprintf "test_conformance_%s" stem

    [<Theory>]
    [<MemberData(nameof(ConformanceTests.GetSpecFiles))>]
    member this.``Run Conformance Test`` (file: string) =
        let testName = ConformanceTests.GetTestName(file)
        let filename = Path.GetFileName(file)
        let options = JsonSerializerOptions()
        options.Converters.Add(NodeConverter())
        options.Converters.Add(ClassItemConverter())
        options.Converters.Add(IROpConverter())
        options.Converters.Add(IRClassItemConverter())

        let json = File.ReadAllText(file)
        use doc = JsonDocument.Parse(json)
        let root = doc.RootElement

        // Output test name for audit visibility (Console) and xUnit logging (output)
        let runMsg = sprintf "=== RUN   %s (%s)" testName filename
        Console.WriteLine(runMsg)
        output.WriteLine(runMsg)

        // Check for error test case
        let mutable expectedErrorElem = Unchecked.defaultof<JsonElement>
        if root.TryGetProperty("expected_error", &expectedErrorElem) then
            // Error test case
            let expectedError = expectedErrorElem.GetString()

            // Only run if input_ast exists
            let mutable inputAstElem = Unchecked.defaultof<JsonElement>
            if root.TryGetProperty("input_ast", &inputAstElem) then
                try
                    let inputAst = JsonSerializer.Deserialize<Node>(inputAstElem.GetRawText(), options)
                    let _ = Compiler.compile inputAst
                    failwithf "Expected error '%s' but compilation succeeded" expectedError
                with
                | _ ->
                    // Expected error caught
                    let passMsg = sprintf "    --- PASS: Caught expected error: %s" expectedError
                    Console.WriteLine(passMsg)
                    output.WriteLine(passMsg)
            else
                // Parser error test: parse input_dsl and verify error + hint
                let mutable inputDslElem = Unchecked.defaultof<JsonElement>
                if root.TryGetProperty("input_dsl", &inputDslElem) then
                    let inputDsl = inputDslElem.GetString()
                    if not (String.IsNullOrEmpty(inputDsl)) then
                        try
                            let _result = Parser.parse inputDsl
                            failwithf "Expected parse error '%s' but parsing succeeded" expectedError
                        with
                        | :? STRlingParseError as parseErr ->
                            // Verify error message contains expected substring
                            if not (parseErr.ErrorMessage.Contains(expectedError)) then
                                failwithf "Error message mismatch.\n  Expected substring: %s\n  Actual: %s" expectedError parseErr.ErrorMessage
                            // Verify hint if expected
                            let mutable expectedHintElem = Unchecked.defaultof<JsonElement>
                            if root.TryGetProperty("expected_hint", &expectedHintElem) then
                                let expectedHint = expectedHintElem.GetString()
                                if not (String.IsNullOrEmpty(expectedHint)) then
                                    match parseErr.Hint with
                                    | Some actualHint ->
                                        if actualHint <> expectedHint then
                                            failwithf "Hint mismatch.\n  Expected: %s\n  Actual: %s" expectedHint actualHint
                                    | None ->
                                        failwithf "Expected hint '%s' but got None" expectedHint
                            let passMsg = sprintf "    --- PASS: Parser error verified: %s" expectedError
                            Console.WriteLine(passMsg)
                            output.WriteLine(passMsg)
                        | ex ->
                            // Non-STRlingParseError, still a pass if it's an error
                            let passMsg = sprintf "    --- PASS: Caught error: %s" (ex.Message)
                            Console.WriteLine(passMsg)
                            output.WriteLine(passMsg)
                    else
                        let passMsg = sprintf "    --- PASS: Parser test (no AST), out of scope"
                        Console.WriteLine(passMsg)
                        output.WriteLine(passMsg)
                else
                    let passMsg = sprintf "    --- PASS: Parser test (no AST), out of scope"
                    Console.WriteLine(passMsg)
                    output.WriteLine(passMsg)

            ()
        else
            // Only run if input_ast exists
            let mutable inputAstElem = Unchecked.defaultof<JsonElement>
            if root.TryGetProperty("input_ast", &inputAstElem) then
                let inputAst = JsonSerializer.Deserialize<Node>(inputAstElem.GetRawText(), options)
                let expectedIr = JsonSerializer.Deserialize<IROp>(root.GetProperty("expected_ir").GetRawText(), options)

                let actualIr = Compiler.compile inputAst

                if actualIr <> expectedIr then
                    let actualJson = JsonSerializer.Serialize(actualIr, options)
                    let expectedJson = JsonSerializer.Serialize(expectedIr, options)
                    failwithf "File: %s\nExpected: %s\nActual:   %s" filename expectedJson actualJson
